src/mlir/verona-mlir.cc

```diff
@@ -92,10 +109,136 @@ namespace
         outputFile = newName.c_str();
       }
     }
+  }
+
+  /// Check the value of command-line arguments, beyond the verification
+  /// provided by cl::opt.
+  ///
+  /// Prints an error message and returns a failure if an invalid combination is
+  /// used.
+  mlir::LogicalResult verifyCommandLine()
+  {
+    if (splitInputFile && inputKind != InputKind::MLIR)
+    {
+      llvm::errs() << "-split-input-file can only be used with MLIR inputs\n";
+      return mlir::failure();
+    }
 
-    // Default grammar
-    // FIXME: Move to llvm::sys::path, but LLVM's GetMainExecutable is horrible
-    grammarFile = path::directory(path::executable()).append("/grammar.peg");
+    if (verifyDiagnostics && inputKind != InputKind::MLIR)
+    {
+      llvm::errs() << "-verify-diagnostics can only be used with MLIR inputs\n";
+      return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult openInput(std::unique_ptr<llvm::MemoryBuffer>* input)
+  {
+    std::string err;
+    *input = mlir::openInputFile(inputFile, &err);
+    if (!*input)
+    {
+      llvm::errs() << err << "\n";
+      return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult openOutput(std::unique_ptr<llvm::ToolOutputFile>* output)
+  {
+    std::string err;
+    *output = mlir::openOutputFile(outputFile, &err);
+    if (!*output)
+    {
+      llvm::errs() << err << "\n";
+      return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult processVeronaInput(llvm::raw_ostream& output)
+  {
+    llvm::ExitOnError check;
+
+    // Parse the file
+    err::Errors errors;
+    module::Passes passes = {sym::build, ref::build, prec::build};
+    auto m = module::build(grammarFile, passes, inputFile, "verona", errors);
+    if (!errors.empty())
+    {
+      llvm::errs() << "ERROR: cannot parse Verona file " << inputFile << "\n"
+                   << errors.to_s() << "\n";
+      return mlir::failure();
+    }
+
+    // Parse AST file into MLIR
+    mlir::verona::Driver driver(optLevel);
+    llvm::Error error = driver.readAST(m->ast);
+    if (!error)
+      error = driver.emitMLIR(output);
+
+    if (error)
+    {
+      driver.dumpMLIR(llvm::errs());
+      logAllUnhandledErrors(std::move(error), llvm::errs());
+      return mlir::failure();
+    }
+    else
```

Comment:
confusing swap of meaning. a simpler sequence would be:
```
error = readAST();
if (!error)
  error = emitMLIR();
if (!error)
  return success;

dumpMLIR()
logErrors()
return failure;
```

same applies to mlirBuffer below